grafanalib/core.py

```diff
@@ -826,8 +1107,49 @@ def to_json_data(self):
             'timepicker': self.timePicker,
             'timezone': self.timezone,
             'version': self.version,
+            'graphTooltip': self.graphTooltip,
         }
 
+    def parse_json_data(data):
+        if '__inputs' in data:
+            data['inputs'] = parse_inputs(data.pop('__inputs'))
+        if 'annotations' in data:
+            data['annotations'] = Annotations.parse_json_data(
+                data['annotations'])
+        if 'templating' in data:
+            data['templating'] = Templating.parse_json_data(data['templating'])
+        if 'rows' in data:
+            data['rows'] = [Row.parse_json_data(row) for row in data['rows']]
+        if 'time' in data:
+            data['time'] = Time.parse_json_data(data['time'])
+        if 'timepicker' in data:
+            data['timePicker'] = TimePicker.parse_json_data(
+                data.pop('timepicker'))
+        if 'links' in data:
+            data['links'] = [DashboardLink.parse_json_data(link)
+                             for link in data['links']]
+
```

Comment:
This pattern appears in a couple of places. I wonder if we can factor something out.

```python
def parse_json_data(data):
    new_data = transform_dict(
        data,
        __inputs=('inputs', parse_inputs),
        annotations=('annotations', Annotations.parse_json_data),
        ...
    )
    return Dashboard(**new_data)

def transform_dict(d, **keys):
    """Apply functions to individual keys of the dict, optionally renaming them.

    :param dict d: Dictionary to be transformed
    :param **keys: Keyword arguments, one for each key in ``d`` that we want to transform. Values should be tuples like ``(new_name, function)``, where ``new_name`` is the new name and ``function`` is applied to the value. If ``d`` doesn't have a key matching ``old_name`` then ``function`` will not get called.
    """
    new_d = dict(d)
    for old_key, (new_key, function) in keys.items():
        if old_key in new_d:
            new_d[new_key] = function(new_d.pop(old_key))
    return new_d
```

What do you think? I think it's possible that we might want the syntax to be `transform_dict(d, new_key=(old_key, function))` rather than the current `transform_dict(d, old_key=(new_key, function))`. If we did that, should also reverse `rename` above.

Reply:
I ended up doing it a bit differently:

```
class dicttransform(
        namedtuple('dicttransform', ['old_key', 'new_key', 'transform'])):
    '''Helper data structure used to simplify the transformations on data when
    parsing the object from JSON. Used in conjunction with `transform_dict`
    :param str old_key: Source key
    :param str new_key: Destination key. If not specified same as `old_key``
    :param function transform: Function used to transform the data, if not
                               specified uses identity function
    Example usage:
    dicttransform('old_key', 'new_key') - renames keys
    dicttransform('old_key', transform=some_function)
     - uses some_function to transform data at `old_key`
    dicttransform('old_key', 'new_key', some_function)
     - use some_function to transform the data and rename the key from
       `old_key` to `new_key`
    '''

    def __new__(cls, old_key, new_key=None, transform=lambda x: x):
        if new_key is None:
            new_key = old_key

        return super(dicttransform, cls).__new__(cls, old_key, new_key,
                                                 transform)


def transform_dict(data, *key_transformations):
    '''
    :param dict data: Dictionary to be transformed
    :param list(dicttransform): Transformation to perform
    '''
    new_data = dict(data)
    for old_key, new_key, transform in key_transformations:
        if old_key in new_data:
            old_value = new_data.pop(old_key)
            new_data[new_key] = transform(old_value)
    return new_data
```

As there is a case for:
1. Renaming the key and keeping the value as it was
2. Processing the value (keeping the key name)
3. Both 1 and 2, process the value and rename the key

Implemented in https://github.com/weaveworks/grafanalib/pull/91/commits/a8bd301a393a81bfbdd2e18a8bae15ef7ab709f6
Looks way cleaner.

What do you think?